compiler/optimizer/VirtualGuardHeadMerger.cpp

```diff
@@ -46,33 +47,96 @@ static bool isMergeableGuard(TR::Node *node)
+      anchored = true;
+      }
+   else
+      {
+      for (int i = 0; i <node->getNumChildren(); i++)
+         anchored |= anchorCommonNodes(comp, node->getChild(i), tt, checklist);
+      }
+
+   return anchored;
+   }
+
+/**
+ * A runtime guard block may have monitor stores and privarg stores along with the guard
+ * it self. This method will rearrange these stores and split the block, managing any
+ * uncommoning necessary for eventual block order.
+ *
+ * The provided block will become the privarg block, containing any privarg stores and additonal
+ * temps for uncommoning. It must be evaluated first. The returned block will contain monitor
+ * stores and the guard. If no split is required, the provided block will be returned.
+ *
+ * @param comp Compilation object
+ * @param block Block to manipulate
+ * @param cfg Current CFG
+ * @return The block containing the guard.
+ */
+static TR::Block* splitRuntimeGuardBlock(TR::Compilation *comp, TR::Block* block, TR::CFG *cfg)
+   {
+   TR::NodeChecklist checklist(comp);
    TR::TreeTop *start = block->getFirstRealTreeTop();
-   TR::TreeTop *end = block->getLastRealTreeTop();
-   bool hasMonitorStore = false;
+   TR::TreeTop *guard = block->getLastRealTreeTop();
    TR::TreeTop *firstPrivArg = NULL;
-   for(TR::TreeTop *tt = start; tt && tt != end; tt = tt->getNextTreeTop())
+   TR::TreeTop *firstMonitor = NULL;
+
+   TR_ASSERT(isMergeableGuard(guard->getNode()), "last node must be guard %p", guard->getNode());
+
+   // Search for privarg and monitor stores, anchoring any commoned loads at the start of the block
+   bool anchoredTemps = false;
+   for (TR::TreeTop *tt = start; tt && tt->getNode()->getOpCodeValue() != TR::BBEnd; tt = tt->getNextTreeTop())
       {
       TR::Node * node = tt->getNode();
+
+      anchoredTemps |= anchorCommonNodes(comp, node, start, checklist);
+
       if (node->getOpCode().hasSymbolReference() && node->getSymbol()->holdsMonitoredObject())
-         hasMonitorStore = true;
-      else if(node->chkIsPrivatizedInlinerArg())
-         {
-         firstPrivArg = tt;
-         // Monitor stores always appear before priv args, so we can safely stop here
-         break;
-         }
+         firstMonitor = firstMonitor == NULL ? tt : firstMonitor;
+      else if (node->chkIsPrivatizedInlinerArg())
+         firstPrivArg = firstPrivArg == NULL ? tt : firstPrivArg;
+      else if (!isMergeableGuard(node))
+         TR_ASSERT(0, "Node other than monitor or privarg store %p before runtime guard", node);
       }
-   if (hasMonitorStore && firstPrivArg)
-      return block->split(firstPrivArg, cfg, true /* fixupCommoning */, false /* copyExceptionSuccessors */);
-   else if (firstPrivArg)
-      return block;
-   else
-      // This block does not contain a priv arg
-      return NULL;
+
+   // If there are privargs and monitors, they must be swapped around, such that all privargs are
```

Comment:
As the monitor stores are before the privarg stores, this will not be the case. It is also not possible to have the inverse, as monitor meta data should not be loaded as arguments, as indicated by a comment in the prior implementation.

It should also be noted that volatile stores are not considered here, as this optimization will see their loads under separate treetops and give in.